File: apps/api-go/internal/infrastructure/container.go
// Package infrastructure wires concrete adapters to the application services.
// It is the ONLY place that imports concrete adapters; the domain and
// application layers depend solely on interfaces. Swapping stub for real
// implementations is a change confined to this file.
package infrastructure

import (
	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/adapter/outbound/stub"
	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/application"
	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/config"
	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/domain/port"
)

// Container holds the singleton adapters shared across requests and builds the
// per-use-case services on demand.
//
// # Pattern: Factory / Composition Root
//
// The Container is the application's composition root: a single factory that
// decides which concrete adapter implements each port. Today every builder
// returns a stub; switching to production means returning the real adapter when
// cfg.UseStubs is false (the //nolint comments mark where they'd plug in).
type Container struct {
	Cfg config.Config

	campaignRepo port.CampaignRepository
	clientRepo   port.ClientRepository
	jobStore     port.JobStore
	adLibrary    port.AdLibrary
	imageGen     port.ImageGenerator
	copyGen      port.CopyGenerator
	storage      port.Storage
}

// NewContainer builds the container and its singleton adapters from config.
func NewContainer(cfg config.Config) *Container {
	c := &Container{Cfg: cfg}
	c.campaignRepo = c.buildCampaignRepo()
	c.clientRepo = c.buildClientRepo()
	c.jobStore = c.buildJobStore()
	c.adLibrary = c.buildAdLibrary()
	c.imageGen = c.buildImageGenerator()
	c.copyGen = c.buildCopyGenerator()
	c.storage = c.buildStorage()
	return c
}

// ── outbound adapter factories ───────────────────────────────────────────────
// Each returns an interface (the port). The concrete type chosen depends on
// cfg.UseStubs. Returning the interface keeps callers oblivious to the choice.

func (c *Container) buildCampaignRepo() port.CampaignRepository {
	// if !c.Cfg.UseStubs { return postgres.NewCampaignRepository(c.Cfg.DatabaseURL) }
	return stub.NewInMemoryCampaignRepository()
}

func (c *Container) buildClientRepo() port.ClientRepository {
	return stub.NewInMemoryClientRepository()
}

func (c *Container) buildJobStore() port.JobStore {
	// if !c.Cfg.UseStubs { return redisadapter.NewJobStore(c.Cfg.RedisURL) }
	return stub.NewInMemoryJobStore()
}

func (c *Container) buildAdLibrary() port.AdLibrary {
	// if !c.Cfg.UseStubs { return meta.NewAdLibrary(c.Cfg.MetaSystemUserToken) }
	return stub.NewAdLibrary()
}

func (c *Container) buildImageGenerator() port.ImageGenerator {
	// if !c.Cfg.UseStubs { return higgsfield.New(...) }
	return stub.NewImageGenerator()
}

func (c *Container) buildCopyGenerator() port.CopyGenerator {
	// if !c.Cfg.UseStubs { return claude.New(c.Cfg.AnthropicAPIKey) }
	return stub.NewCopyGenerator()
}

func (c *Container) buildStorage() port.Storage {
	// if !c.Cfg.UseStubs { return gcs.New(c.Cfg.GCSBucketName) }
	return stub.NewStorage()
}

// ── inbound service factories ────────────────────────────────────────────────

// CampaignService builds a campaign service bound to a client's ad account.
// In production the AdPlatform (Meta CLI) is created per request because it
// carries the client's AD_ACCOUNT_ID; in stub mode it's a shared fake.
func (c *Container) CampaignService(adAccountID string) *application.CampaignService {
	var adPlatform port.AdPlatform = stub.NewAdPlatform()
	// if !c.Cfg.UseStubs { adPlatform = meta.NewCLIAdapter(c.Cfg.MetaSystemUserToken, adAccountID) }
	_ = adAccountID
	return application.NewCampaignService(adPlatform, c.campaignRepo, c.jobStore)
}

// InspirationService builds the AI pipeline service from the shared adapters.
func (c *Container) InspirationService() *application.InspirationService {
	return application.NewInspirationService(c.adLibrary, c.imageGen, c.copyGen, c.storage, c.jobStore)
}

// ClientRepo exposes the client repository for read-only handlers.
func (c *Container) ClientRepo() port.ClientRepository { return c.clientRepo }

File: apps/api/infrastructure/container.py
"""Dependency injection: the ONLY place that constructs concrete adapters.

Picks stub or real adapters based on settings.USE_STUBS. The domain and
application layers never import anything from here.
"""
from __future__ import annotations

from functools import lru_cache

from application.campaign_service import CampaignService
from application.inspiration_service import InspirationService
from config import Settings, get_settings
from domain.ports.outbound.ad_library import IAdLibrary
from domain.ports.outbound.campaign_repository import ICampaignRepository
from domain.ports.outbound.client_repository import IClientRepository
from domain.ports.outbound.copy_generator import ICopyGenerator
from domain.ports.outbound.image_generator import IImageGenerator
from domain.ports.outbound.job_store import IJobStore


class Container:
    """Holds singletons shared across requests/workers."""

    def __init__(self, settings: Settings | None = None):
        self.settings = settings or get_settings()
        self.campaign_repo: ICampaignRepository = self._build_campaign_repo()
        self.client_repo: IClientRepository = self._build_client_repo()
        self.job_store: IJobStore = self._build_job_store()
        self.ad_library: IAdLibrary = self._build_ad_library()
        self.image_generator: IImageGenerator = self._build_image_generator()
        self.copy_generator: ICopyGenerator = self._build_copy_generator()

    # ── outbound adapters ────────────────────────────────────────────────
    def _build_campaign_repo(self) -> ICampaignRepository:
        if self.settings.USE_STUBS:
            from adapters.outbound.stubs import InMemoryCampaignRepository

            return InMemoryCampaignRepository()
        from adapters.outbound.postgres_repository import PostgresCampaignRepository

        return PostgresCampaignRepository(self.settings.DATABASE_URL)

    def _build_client_repo(self) -> IClientRepository:
        # Clients are reference data; seeded in-memory for the scaffold.
        from adapters.outbound.stubs import InMemoryClientRepository

        return InMemoryClientRepository()

    def _build_job_store(self) -> IJobStore:
        if self.settings.USE_STUBS:
            from adapters.outbound.stubs import InMemoryJobStore

            return InMemoryJobStore()
        from adapters.outbound.redis_job_store import RedisJobStore

        return RedisJobStore(self.settings.REDIS_URL)

    def _build_ad_library(self) -> IAdLibrary:
        if self.settings.USE_STUBS:
            from adapters.outbound.stubs import StubAdLibrary

            return StubAdLibrary()
        from adapters.outbound.ad_library_adapter import AdLibraryAdapter

        return AdLibraryAdapter(self.settings.META_SYSTEM_USER_TOKEN)

    def _build_image_generator(self) -> IImageGenerator:
        if self.settings.USE_STUBS:
            from adapters.outbound.stubs import StubImageGenerator

            return StubImageGenerator()
        from adapters.outbound.higgsfield_adapter import HiggsfieldAdapter

        return HiggsfieldAdapter(self.settings.HF_API_KEY, self.settings.HF_API_SECRET)

    def _build_copy_generator(self) -> ICopyGenerator:
        if self.settings.USE_STUBS:
            from adapters.outbound.stubs import StubCopyGenerator

            return StubCopyGenerator()
        from adapters.outbound.claude_adapter import ClaudeAdapter

        return ClaudeAdapter(self.settings.ANTHROPIC_API_KEY)

    # ── inbound services ─────────────────────────────────────────────────
    def campaign_service(self, ad_account_id: str) -> CampaignService:
        """Per-request: the CLI adapter is bound to the client's ad_account_id."""
        if self.settings.USE_STUBS:
            from adapters.outbound.stubs import StubAdPlatform

            ad_platform = StubAdPlatform()
        else:
            from adapters.outbound.meta_cli_adapter import MetaCLIAdapter

            ad_platform = MetaCLIAdapter(
                access_token=self.settings.META_SYSTEM_USER_TOKEN,
                ad_account_id=ad_account_id,
            )
        return CampaignService(
            ad_platform=ad_platform,
            campaign_repo=self.campaign_repo,
            job_store=self.job_store,
        )

    def inspiration_service(self) -> InspirationService:
        return InspirationService(
            ad_library=self.ad_library,
            image_generator=self.image_generator,
            copy_generator=self.copy_generator,
            job_store=self.job_store,
        )


@lru_cache
def get_container() -> Container:
    return Container()
